README.md
# consul-kv-monitor

[![NPM version](https://img.shields.io/npm/v/consul-kv-monitor.svg)](https://www.npmjs.com/package/consul-kv-monitor)
[![Release Status](https://github.com/LCMApps/consul-kv-monitor/workflows/NPM%20Release/badge.svg)](https://github.com/LCMApps/consul-kv-monitor/releases)
[![Build Status](https://travis-ci.org/LCMApps/consul-kv-monitor.svg?branch=master)](https://travis-ci.org/LCMApps/consul-kv-monitor)
[![Coverage Status](https://coveralls.io/repos/github/LCMApps/consul-kv-monitor/badge.svg?branch=master)](https://coveralls.io/github/LCMApps/consul-kv-monitor?branch=master)

Module monitors consul key-value store and notifies about changes of a value or a set of values. It relies on
blocking queries and detects changes as soon as consul agent receives them. This module uses
[`consul`](https://github.com/silas/node-consul) module that must be installed too.

Please, check the full documentation below.

**Table of Contents**

* [Installation and Usage](#installation)
* [Monitor config](#monitor-config)
* [Listening of changes](#listening-of-changes)
* [Failures](#failures)

# <a name="installation"></a>Installation and Usage

Using npm:
```shell
$ npm install --save consul-kv-monitor consul
```

Using yarn:
```shell
$ yarn add consul-kv-monitor consul
```

After you've installed both `consul` and `consul-kv-monitor`, set it up and start the service in your code.

```js
const consul = require('consul');
const consulClient = consul({
    host: '127.0.0.1',
    port: 8500
});

const {ConsulKvMonitor, Errors} = require('consul-kv-monitor');

const monitorConfig = {
    keysPrefix: 'lcm',
    timeoutMsec: 1000,
    json: false
};

const monitor = new ConsulKvMonitor(monitorConfig, consulClient);

function printData(consulKvData) {
    const keys = consulKvData.getKeys();
    keys.forEach(key => {
        let value = consulKvData.getValue(key);
        if (value === Object(value)) {
            // we received json decoded data, so just stringifying to pretty output in this example
            value = JSON.stringify(value);
        }

        const metadata = consulKvData.getMetadata(key);
        console.log(`${key} => ${value}, metadata: ${JSON.stringify(metadata)}`);
    });
}

monitor.on('changed', (kvData) => {
    console.log('Some data has changed');
    console.log('Response headers:', JSON.stringify(monitor.getConsulHeaders()));
    printData(kvData);
    console.log();
});

monitor.on('error', (error) => {
    if (error instanceof Errors.InvalidDataError) {
        console.log(`Oh, key "${error.extra.key}" value can not be decoded as JSON, actual value is "${error.extra.value}"`);
        console.log('And here is a raw error');
        console.log(error);
    } else {
        console.log(`Error occured, class: ${error.name}`);
        console.log('And here is a raw error');
        console.log(error);
    }
});

monitor.start()
    .then(initialData => {
        console.log('Keys discovered on start:');
        console.log('Response headers:', JSON.stringify(monitor.getConsulHeaders()));
        printData(initialData);
        console.log();
    })
    .catch(err => {
        console.log(err instanceof Errors.WatchTimeoutError);
        console.error(err);
        process.exit(1);
    });
```

Lets add some data to kv:

```shell
$ consul kv put lcm/server-1 10.0.0.2
Success! Data written to: lcm/server-1
```

Then start this example. Right after initialization we may discover our key

```shell
$ node index.js
Keys discovered on start:
Response headers: {"x-consul-index":"337151414","x-consul-knownleader":"true","x-consul-lastcontact":"0"}
lcm/server-1 => 10.0.0.2, metadata: {"LockIndex":0,"Key":"lcm/server-1","Flags":0,"Value":"10.0.0.2","CreateIndex":337151414,"ModifyIndex":337151414}
```

Lets add one more key to subpath under monitoring while example is running.

```shell
$ consul kv put lcm/server-2 10.0.0.3
Success! Data written to: lcm/server-2
```

Monitor detects this change:

```shell
Some data has changed
Response headers: {"x-consul-index":"337156828","x-consul-knownleader":"true","x-consul-lastcontact":"0"}
lcm/server-1 => 10.0.0.2, metadata: {"LockIndex":0,"Key":"lcm/server-1","Flags":0,"Value":"10.0.0.2","CreateIndex":337151414,"ModifyIndex":337151414}
lcm/server-2 => 10.0.0.3, metadata: {"LockIndex":0,"Key":"lcm/server-2","Flags":0,"Value":"10.0.0.3","CreateIndex":337156828,"ModifyIndex":337156828}
```

Delete `lcm/server-1` and update `lcm/server-2`:

```shell
$ consul kv delete lcm/server-1
Success! Deleted key: lcm/server-1

$ consul kv put lcm/server-2 10.0.0.4
Success! Data written to: lcm/server-2
```

Monitor detects our changes:

```shell

Some data has changed
Response headers: {"x-consul-index":"337158175","x-consul-knownleader":"true","x-consul-lastcontact":"0"}
lcm/server-2 => 10.0.0.3, metadata: {"LockIndex":0,"Key":"lcm/server-2","Flags":0,"Value":"10.0.0.3","CreateIndex":337156828,"ModifyIndex":337156828}

Some data has changed
Response headers: {"x-consul-index":"337158295","x-consul-knownleader":"true","x-consul-lastcontact":"0"}
lcm/server-2 => 10.0.0.4, metadata: {"LockIndex":0,"Key":"lcm/server-2","Flags":0,"Value":"10.0.0.4","CreateIndex":337156828,"ModifyIndex":337158295}
```

And finally delete all keys:

```shell
$ consul kv delete lcm/server-2
Success! Deleted key: lcm/server-2
```

Monitor emits `changed` event with an object without any keys.

```shell
Some data has changed
Response headers: {"x-consul-index":"337159004","x-consul-knownleader":"true","x-consul-lastcontact":"0"}
```

# <a name="monitor-config"></a>Monitor Config

There are few options available for the config object:
* `keysPrefix` (String): a path to specific key or path to set of keys, monitor always use
[`recurse`](https://github.com/silas/node-consul#consulkvgetoptions-callback) option of the `consul` client
* `timeoutMsec` (Number, optional, deafult: 5000): number of milliseconds to wait initial response from consul
* `json` (Boolean, optional, default: false): if `true` tries to decode json object from stringified value from consul,
make sense only if you store values as stringified objects


Example,

```js
const monitorConfig = {
    keysPrefix: 'path/to/key',
    timeoutMsec: 1000,
    json: true
};

const monitor = new ConsulKvMonitor(monitorConfig, consulClient);
```

Constructor throws error `TypeError` if invalid values passes. 

# <a name="start-and-stop"></a>Start and stop

### `ConsulKvMonitor.start()`

To start monitoring just call `start` method of the monitor.

```js
Errors = require('consul-kv-monitor').Errors;

try {
    const initialData = await monitor.start();
} catch (err) {
    console.log(err instanceof Errors.WatchTimeoutError);
}
```

`start` method returns promise that may be resolved with values under monitoring or be rejected.

Promise rejects with one of the following errors (all of them are in `Errors` set):
* `AlreadyInitializedError` if service is already started.
* `WatchTimeoutError` if either initial data nor error received for `timeoutMsec` msec
* `WatchError` on error from `consul` underlying method

Promise resolved only once. Rejection of promise means that watcher was stopped and no retries will be done.
To receive updates you may [add listeners](#listening-of-changes).

After successful start monitor never gives up and tries to reconnect even after failures. Of course, it [notifies
about failures](#failures).

### `ConsulKvMonitor.stop()`

To stop monitor just call `stop` method. It returns monitor object itself.

### `ConsulKvMonitor.isInitialized()`

This method returns `false` in the following scenarios:
* before start of monitoring
* [after failure](#failures) and till recovering from that failure
* after stop of monitor

After successful start it returns `true` till one of the event described above occurs.

# <a name="listening-of-changes"></a>Listening of changes

Monitor emits `changed` event if values in consul or data's metadata changes. It's possible to catch `changed` event
with a data without any modifications comparing to previous catch. It happens because consul's fields like
`ModifyIndex` or `LockIndex` may change without modification of data.

So, it's a good idea to compare actual changes.

Anyway, monitor emits instance of `ConsulKvData` class that has the following methods:
* `ConsulKvData.hasKey(key)`: checks presence of the given key in a set of monitored keys and returns `Boolean`
* `ConsulKvData.getKeys()`: returns an array of keys
* `ConsulKvData.getValue(key)`: returns [decoded json data](#monitor-config) or raw string data, if key is absent
returns `undefined`
* `ConsulKvData.getMetadata(key)`: returns all fields received from consul

Also, you may get `ConsulKvData` object by direct call to monitor at any moment after start:

```js
const keys = monitor.getData().getKeys();
```

`monitor.getData()` returns `ConsulKvData` object even if there are no keys at all.

# <a name="failures"></a>Failures

While monitor listens for changes it can lose connection with consul agent or consul agent can lose connection with a
master server, so keys under monitoring may become inconsistent. Monitor will run normally and will try to recover
as soon as possible using reconnection with backoff. But business logic may require to detect such situations and make
some actions. At [LCMApps](https://github.com/LCMApps) we stop processing requests while kv connection is unhealthy.

So, monitor emits `unhealthy` event if it detects failure. You may still get last seen values from KV or consul headers
but always remember that requested data may be stale.

If a unhealthy state is caused by consul error then monitor emits `error` event with `WatchError` instance right
after `unhealthy` event.

After successful recovery to healthy state monitor emits `healthy` event and may emit `changed` event with updated
keys.  

At any time you may get health status running method explicitly:

```js
monitor.isWatchHealthy();
```

Method returns `true` or `false`.

Also there is special case when monitor emits `error` events. If you pass `json: true` option in constructor monitor
tries to decode string in value field as a JSON. But if value is not JSON monitor can't decode value and emits
`error` event passing object of `InvalidDataError` class. For example, if consul KV stores 3 values not in
stringified JSON format monitor emits 3 errors.

Let's use an example from the start of this readme, but with `json: true` option.


Set invalid JSON value

```shell
$ consul kv put lcm/server-2 '{"a":1,"b"}'
Success! Data written to: lcm/server-2
```

Monitor emits `error`

```shell
Some data has changed
Response headers: {"x-consul-index":"337175801","x-consul-knownleader":"true","x-consul-lastcontact":"0"}

Oh, key "lcm/server-2" value can not be decoded as JSON, actual value is "{"a":1,"b"}"
And here is a raw error
{ InvalidDataError: Invalid JSON of Value field of KV is received from consul, record will be skipped
    at validRecords.forEach.record (/home/vss-services-rel/mt/node_modules/consul-kv-monitor/src/Factory.js:31:31)
    at Array.forEach (<anonymous>)
    at Object.buildConsulKvData (/home/vss-services-rel/mt/node_modules/consul-kv-monitor/src/Factory.js:22:18)
    at ConsulKvMonitor._onWatcherChange (/home/vss-services-rel/mt/node_modules/consul-kv-monitor/src/ConsulKvMonitor.js:274:50)
    at emitTwo (events.js:126:13)
    at Watch.emit (events.js:214:7)
    at /home/vss-services-rel/mt/node_modules/consul/lib/watch.js:179:14
    at Consul.<anonymous> (/home/vss-services-rel/mt/node_modules/consul/lib/kv.js:71:5)
    at next (/home/vss-services-rel/mt/node_modules/papi/lib/client.js:313:25)
    at IncomingMessage.<anonymous> (/home/vss-services-rel/mt/node_modules/papi/lib/client.js:611:7)
  extra: { key: 'lcm/server-2', value: '{"a":1,"b"}' },
  name: 'InvalidDataError' }
```

Set correct JSON

```shell
$ consul kv put lcm/server-2 '{"a":1,"b":[]}'
Success! Data written to: lcm/server-2
```

Monitor sees changes

```shell
Some data has changed
Response headers: {"x-consul-index":"337176075","x-consul-knownleader":"true","x-consul-lastcontact":"0"}
lcm/server-2 => {"a":1,"b":[]}, metadata: {"LockIndex":0,"Key":"lcm/server-2","Flags":0,"Value":"{\"a\":1,\"b\":[]}","CreateIndex":337171339,"ModifyIndex":337176075}
```


CHANGELOG.md
# Changelog

### 2.0.3

- Dependencies was bumped

### 2.0.2

BUG FIXES:

- Potential bug with rejection of promise
[[GH-14](https://github.com/LCMApps/consul-kv-monitor/issues/14)]
- Does not emit "healthy" event in method
[[GH-15](https://github.com/LCMApps/consul-kv-monitor/issues/15)]
- JSDoc fixes

### 2.0.1

IMPROVEMENTS:

- Removing module "asyncawait" and support of Nodejs 6

### 2.0.0

IMPROVEMENTS:

- Add forwarding headers X-Consul-*
- Fix bug with default timeout for Consul Watch
- Add to consul-kv-monitor module functionality of an auto-reconnect to Сonsul
- Support auto-reconnect to Consul 
- Removed `emergencyStop` event from `ServiceInstancesMonitor`
- Added `healthy` and `unhealthy` events to `ServiceInstancesMonitor`


tests/Functional/nock.data.js
const loadData1 = {
    status: 'OK',
    pid: 100,
    mem: {
        total: 12813,
        free: 11786
    },
    cpu: {
        usage: 0.72,
        count: 16
    }
};

module.exports = {
    loadData1: loadData1,
    firstResponseHeaders: {
        'X-Consul-Index': '313984',
        'X-Consul-Knownleader': 'true',
        'X-Consul-Lastcontact': '0'
    },
    firstResponseBody: [
        {
            'CreateIndex': 351643,
            'Flags': 0,
            'Key': 'mediaservermetrics/192.168.101.5',
            'LockIndex': 0,
            'ModifyIndex': 906432,
            'Value': '"dGVzdA=="'
        },
        {
            'CreateIndex': 350295,
            'Flags': 0,
            'Key': 'mediaservermetrics/192.168.101.6',
            'LockIndex': 0,
            'ModifyIndex': 906434,
            'Value': '"dGVzdA=="'
        }
    ]
};



src/Factory.js
'use strict';

const InvalidDataError = require('./Error').InvalidDataError;
const ConsulKvData = require('./ConsulKvData');
const ConsulResponseValidator = require('./ConsulResponseValidator');

/**
 * Function receives path of the key which was read and body of response. Result is ConsulKvData and list of errors.
 *
 * @param {Array|undefined} data
 * @param {Boolean} json
 * @returns {{consulKvData: ConsulKvData, errors: Array}}
 */
function buildConsulKvData(data, json = false) {
    if (data === undefined) {
        return {consulKvData: new ConsulKvData([]), errors: []};
    }

    const {validRecords, errors} = ConsulResponseValidator.filterValidKvRecords(data);
    const records = [];

    validRecords.forEach(record => {
        const body = {
            metaData: record
        };

        if (json) {
            try {
                body.value = JSON.parse(record.Value);
            } catch (err) {
                const error = new InvalidDataError(
                    'Invalid JSON of Value field of KV is received from consul, record will be skipped',
                    {
                        key: record.Key,
                        value: record.Value
                    }
                );

                return errors.push(error);
            }
        } else {
            body.value = record.Value;
        }

        records.push([record.Key, body]);
    });

    const consulKvData = new ConsulKvData(records);
    return {consulKvData, errors};
}

module.exports = {
    buildConsulKvData
};


src/ConsulKvData.js
'use strict';

class ConsulKvData {
    /**
     * Set records to the map
     *
     * @param {Array} [records]
     */
    constructor(records) {
        this._kvMap = new Map(records);
    }

    /**
     * @param {String} key
     * @returns {Boolean}
     */
    hasKey(key) {
        return this._kvMap.has(key);
    }

    /**
     * Get all keys from map
     *
     * @returns {String[]}
     */
    getKeys() {
        return Array.from(this._kvMap.keys());
    }

    /**
     * Get saved value from KV
     *
     * @param {String} key
     * @returns {*}
     */
    getValue(key) {
        if (!this._kvMap.has(key)) {
            return undefined;
        }

        return this._kvMap.get(key).value;
    }

    /**
     * Get metadata of record
     *
     * @param {String} key
     * @returns {*}
     */
    getMetadata(key) {
        if (!this._kvMap.has(key)) {
            return undefined;
        }

        return this._kvMap.get(key).metaData;
    }
}

module.exports = ConsulKvData;


src/ConsulResponseValidator.js
'use strict';

const _ = require('lodash');
const InvalidDataError = require('./Error').InvalidDataError;

/**
 * Filters records that have valid format of data. Records that have invalid format will be not be
 * returned back to caller and `errors` array will contain information about format mismatch.
 *
 * Function checks presence of following properties:
 *   - `nodes[n].Key`
 *   - `nodes[n].Value`
 *   - `nodes[n].CreateIndex`
 *   - `nodes[n].Flags`
 *   - `nodes[n].LockIndex`
 *   - `nodes[n].ModifyIndex`
 *
 * @param {*} records - data received from `consul.kv`
 * @returns {{validRecords: Object[], errors: InvalidDataError[]}}
 */
function filterValidKvRecords(records) {
    const data = {validRecords: [], errors: []};

    if (!_.isArray(records)) {
        data.errors.push(new InvalidDataError('Invalid format of data received from consul KV', {records}));

        return data;
    }

    if (_.isEmpty(records)) {
        return data;
    }

    records.forEach(record => {
        if (
            !_.isObject(record) || !_.has(record, 'Key') || !_.has(record, 'Value') ||
            !_.has(record, 'CreateIndex') || !_.has(record, 'Flags') || !_.has(record, 'LockIndex') ||
            !_.has(record, 'ModifyIndex')
        ) {
            data.errors.push(new InvalidDataError('Invalid format of record data received from consul KV', {record}));

            return;
        }


        data.validRecords.push(record);
    });

    return data;
}


module.exports = {
    filterValidKvRecords
};


src/ConsulKvMonitor.js
'use strict';

const EventEmitter = require('events');
const _ = require('lodash');

const AlreadyInitializedError = require('./Error').AlreadyInitializedError;
const WatchError = require('./Error').WatchError;
const WatchTimeoutError = require('./Error').WatchTimeoutError;
const kvFactory = require('./Factory');
const ConsulKvData = require('./ConsulKvData');

const DEFAULT_TIMEOUT_MSEC = 5000;
const HEALTH_FALLBACK_INTERVAL_MSEC = 1000;
const DEFAULT_RETRY_START_SERVICE_TIMEOUT_MSEC = 1000;
const X_CONSUL_HEADERS = ['x-consul-index', 'x-consul-knownleader', 'x-consul-lastcontact'];

/**
 * @emits ConsulKvMonitor#changed
 * @emits ConsulKvMonitor#error
 * @emits ConsulKvMonitor#healthy
 * @emits ConsulKvMonitor#unhealthy
 */
class ConsulKvMonitor extends EventEmitter {
    /**
     * @param {Object} options
     * @param {String} options.keysPrefix - namespace in consul to monitor
     * @param {Boolean} options.json - namespace in consul to monitor
     * @param {Number} [options.timeoutMsec=5000] - connection timeout to consul
     * @param {Consul} consul
     * @throws {TypeError} On invalid options format
     * @public
     */
    constructor(options, consul) {
        super();

        if (!_.isPlainObject(options)) {
            throw new TypeError('options must be an object');
        }

        if (!_.has(options, 'keysPrefix') || !_.isString(options.keysPrefix) || _.isEmpty(options.keysPrefix)) {
            throw new TypeError('options.keysPrefix must be set and be a non-empty string');
        }

        if (!_.has(options, 'timeoutMsec')) {
            this._timeoutMsec = DEFAULT_TIMEOUT_MSEC;
        } else {
            if (!_.isSafeInteger(options.timeoutMsec) || options.timeoutMsec <= 0) {
                throw new TypeError('options.timeoutMsec must be a positive integer if set');
            }

            this._timeoutMsec = options.timeoutMsec;
        }

        // duck typing check
        if (!_.isObject(consul) || !_.isFunction(consul.watch) ||
            !_.isObject(consul.kv) || !_.isFunction(consul.kv.get)
        ) {
            throw new TypeError('consul argument does not look like Consul object');
        }

        this._keysPrefix = options.keysPrefix;
        this._json = options.json;
        this._initialized = false;

        this._consul = consul;

        this._consulKvData = new ConsulKvData();
        this._consulHeaders = {};

        this._onWatcherChange = this._onWatcherChange.bind(this);
        this._onWatcherError = this._onWatcherError.bind(this);
        this._onWatcherEnd = this._onWatcherEnd.bind(this);
        this._retryStartService = this._retryStartService.bind(this);

        this._watchKvChange = null;
        this._setWatchUnhealthy();
        this._setUninitialized();

        this._fallbackToWatchHealthyInterval = null;
        this._retryTimer = null;
    }

    isWatchHealthy() {
        return this._isWatchHealthy;
    }

    _setWatchHealthy() {
        this._isWatchHealthy = true;
    }

    _setWatchUnhealthy() {
        this._isWatchHealthy = false;
    }

    isInitialized() {
        return this._initialized;
    }

    _setInitialized() {
        this._initialized = true;
    }

    _setUninitialized() {
        this._initialized = false;
    }

    _isWatcherRegistered() {
        return this._watchKvChange !== null;
    }

    /**
     * @returns {ConsulKvData}
     */
    getData() {
        return this._consulKvData;
    }

    /**
     * @returns {Object}
     */
    getConsulHeaders() {
        return this._consulHeaders;
    }

    /**
     * Starts service and resolves promise with initial list of KV data.
     *
     * Listens for changes after successful resolve.
     *
     * Promise will be rejected with:
     *   `AlreadyInitializedError` if service is already started.
     *   `WatchTimeoutError` if either initial data nor error received for 5000 msec
     *   `WatchError` on error from `consul` underlying method
     *
     * Rejection of promise means that watcher was stopped and no retries will be done.
     *
     * @returns {Promise.<Array>|Promise.<AlreadyInitializedError>|Promise.<WatchError>|Promise.<WatchTimeoutError>}
     * @public
     */
    start() {
        if (this._isWatcherRegistered()) {
            return Promise.reject(new AlreadyInitializedError('Service is already started'));
        }

        return this._registerWatcherAndWaitForInitialKvData()
            .then(kvData => {
                this._watchKvChange.on('change', this._onWatcherChange);
                this._watchKvChange.on('error', this._onWatcherError);
                this._watchKvChange.on('end', this._onWatcherEnd);

                this._setInitialized();
                this._setWatchHealthy();

                this._consulKvData = kvData;
                return kvData;
            });
    }

    /**
     * Stops service even if it is not started yet. Monitor becomes `uninitialized` and `unhealthy`.
     *
     * Does not listened for changes after execution.
     *
     * @returns {ConsulKvMonitor}
     * @public
     */
    stop() {
        if (this._retryTimer !== null) {
            clearTimeout(this._retryTimer);
            this._retryTimer = null;
        }

        if (!this._isWatcherRegistered()) {
            return this;
        }

        // we need to remove listener to prevent emitting of `end` event after stop of watcher
        this._watchKvChange.removeListener('end', this._onWatcherEnd);
        this._watchKvChange.end();
        this._watchKvChange = null;
        this._unsetFallbackToWatchHealthy();
        this._setUninitialized();
        this._setWatchUnhealthy();
        return this;
    }

    /**
     * Registers `consul.watch` and assigns watcher to `this._watchKvChange` and waits for the
     * first successful response from consul with list of KV data.
     * On successful response resolves promise with array of kv records (it may be empty). Method doesn't
     * add listener for `change` event.
     *
     * Promise will be rejected with:
     *   `AlreadyInitializedError` if another `consul.watch` execution is found.
     *   `WatchTimeoutError` if either initial data nor error received for 5000 msec
     *   `WatchError` on error from `consul` underlying method
     *
     * Rejection of promise means that watch was stopped and `this._watchKvChange` was cleared.
     *
     * @returns {Promise.<Array>|Promise.<AlreadyInitializedError>|Promise.<WatchError>|Promise.<WatchTimeoutError>}
     * @private
     */
    _registerWatcherAndWaitForInitialKvData() {
        return new Promise((resolve, reject) => {
            this._watchKvChange = this._consul.watch({
                method: this._consul.kv.get,
                options: {
                    key: this._keysPrefix,
                    recurse: true,
                    wait: '60s',
                },
            });

            const firstChange = (data, response) => {
                this._watchKvChange.removeListener('error', firstError);
                clearTimeout(timerId);

                const {consulKvData, errors} = kvFactory.buildConsulKvData(data, this._json);

                for (const headerName of X_CONSUL_HEADERS) {
                    this._consulHeaders[headerName] = response.headers[headerName];
                }

                if (!_.isEmpty(errors)) {
                    this._emitFactoryErrors(errors);
                }

                resolve(consulKvData);
            };

            const firstError = (err) => {
                this._watchKvChange.removeListener('change', firstChange);
                this._watchKvChange.end();
                this._watchKvChange = null;
                clearTimeout(timerId);
                reject(new WatchError(err.message, {err}));
            };

            const timerId = setTimeout(() => {
                this._watchKvChange.removeListener('error', firstError);
                this._watchKvChange.removeListener('change', firstChange);
                this._watchKvChange.end();
                this._watchKvChange = null;
                reject(new WatchTimeoutError('Initial consul watch request was timed out'));
            }, this._timeoutMsec);

            this._watchKvChange.once('change', firstChange);
            this._watchKvChange.once('error', firstError);
        });
    }

    /**
     * This method receives list of key-value records sent by `consul.watch` in `consul` format. Performs
     * parsing of response.
     *
     * If service was unhealthy, it becomes healthy.
     *
     * @param {Array} data - list of key-value records after some changes
     * @param {IncomingMessage} response - response from Consul
     * @emits ConsulKvMonitor#changed actual array of key-value records
     * @private
     */
    _onWatcherChange(data, response) {
        let isHealthyStateChanged = false;
        if (!this.isWatchHealthy()) {
            this._setWatchHealthy();
            isHealthyStateChanged = true;
        }

        const {consulKvData, errors} = kvFactory.buildConsulKvData(data, this._json);

        this._consulKvData = consulKvData;
        for (const headerName of X_CONSUL_HEADERS) {
            this._consulHeaders[headerName] = response.headers[headerName];
        }
        if (isHealthyStateChanged) {
            this.emit('healthy');
        }
        this.emit('changed', consulKvData);

        if (!_.isEmpty(errors)) {
            this._emitFactoryErrors(errors);
        }
    }

    /**
     * This method receives an Error from Consul.
     *
     * If service was healthy, it becomes unhealthy.
     *
     * @param {Error} err
     * @emits ConsulKvMonitor#error `WatchError` error
     * @private
     */
    _onWatcherError(err) {
        this._unsetFallbackToWatchHealthy();

        if (this.isWatchHealthy()) {
            this._setWatchUnhealthy();
            this.emit('unhealthy');
        }

        this._setFallbackToWatchHealthy();

        this.emit('error', new WatchError(err.message, {err}));
    }

    /**
     * This method is called when connection with Consul agent was refused.
     *
     * Monitor becomes `uninitialized` and `unhealthy`.
     *
     * @emits ConsulKvMonitor#unhealthy
     * @private
     */
    async _onWatcherEnd() {
        this._unsetFallbackToWatchHealthy();
        this._setUninitialized();
        if (this.isWatchHealthy()) {
            this._setWatchUnhealthy();
            this.emit('unhealthy');
        }
        this._watchKvChange = null;
        await this._retryStartService();
    }

    /**
     * This method receives a list of errors and emits them.
     *
     * @param {Error[]} errors
     * @emits ConsulKvMonitor#error
     * @private
     */
    _emitFactoryErrors(errors) {
        setImmediate(() => {
            errors.forEach(error => this.emit('error', error));
        });
    }

    /**
     * @private
     */
    _setFallbackToWatchHealthy() {
        if (this._fallbackToWatchHealthyInterval) {
            this._unsetFallbackToWatchHealthy();
        }

        const initialUpdateTime = this._watchKvChange.updateTime();

        this._fallbackToWatchHealthyInterval = setInterval(() => {
            if (this.isWatchHealthy()) {

                // watcher is currently ends or becomes `healthy`, unset fallback interval
                this._unsetFallbackToWatchHealthy();

                return;
            }

            const lastUpdateTime = this._watchKvChange.updateTime();

            if (initialUpdateTime !== lastUpdateTime) {
                this._unsetFallbackToWatchHealthy();

                this._setWatchHealthy();
                this.emit('healthy');
            }

        }, HEALTH_FALLBACK_INTERVAL_MSEC);
    }

    /**
     * @private
     */
    _unsetFallbackToWatchHealthy() {
        clearInterval(this._fallbackToWatchHealthyInterval);

        this._fallbackToWatchHealthyInterval = null;
    }

    async _retryStartService() {
        try {
            const consulKvData = await this.start();
            this._consulKvData = consulKvData;

            this.emit('healthy');
            this.emit('changed', consulKvData);
        } catch (err) {
            setImmediate(() => this.emit('error', err));

            this._retryTimer = setTimeout(this._retryStartService, DEFAULT_RETRY_START_SERVICE_TIMEOUT_MSEC);
        }
    }

}

module.exports = ConsulKvMonitor;
